File: tests/test_vec_float.c
/* Copyright (C) 2015, WSID   */

/* This file is part of Crank System.
 *
 *  Crank System is free software: you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation, either
 *  version 3 of the License, or (at your option) any later version.
 *
 *  Crank System is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with Crank System.
 *  If not, see <http://www.gnu.org/licenses/>.
 */

#include <glib.h>
#include "crankbase.h"

void	test_assert_float_real (	const gchar*	file,
								const guint		line,
								const gchar*	func,
								const gchar*	a_str,
								const gchar*	b_str,
								gfloat			a,
								gfloat			b)
{
	if ((a < b - 0.0001f) || (b + 0.0001f < a)) {
		g_printerr ("%s:%u:%s: assertion failed: (%s == %s): (%g == %g) (with diff of %g)",
				file, line, func,
				a_str, b_str,
				a, b, b - a );
	}
}

#define test_assert_float(a, b) \
	test_assert_float_real(__FILE__, __LINE__, G_STRFUNC, \
			G_STRINGIFY(a), G_STRINGIFY(b), \
			a, b)






void	test_2_get (void)
{
	CrankVecFloat2	a = {3.0f, 4.0f};
	
	test_assert_float (crank_vec_float2_get (&a, 0), 3.0f);
	test_assert_float (crank_vec_float2_get (&a, 1), 4.0f);
}


void	test_2_equal (void)
{
	CrankVecFloat2	a = {(gfloat) g_test_rand_double (), (gfloat) g_test_rand_double ()};
	CrankVecFloat2	b = {(gfloat) g_test_rand_double (), (gfloat) g_test_rand_double ()};
	CrankVecFloat2	c = {a.x, a.y};
	
	g_assert (! crank_vec_float2_equal (&a, &b));
	g_assert (  crank_vec_float2_equal (&a, &c));
}


void	test_2_to_string (void)
{
	CrankVecFloat2	a = {3.0f, 4.0f};
	gchar*			astr = crank_vec_float2_to_string (&a);
	
	g_assert_cmpstr (astr, ==, "(3, 4)");
}


void	test_2_magn (void)
{
	CrankVecFloat2	a = {3.0f, 4.0f};
	
	test_assert_float (crank_vec_float2_get_magn_sq (&a), 25.0f);
	test_assert_float (crank_vec_float2_get_magn (&a), 5.0f);
}


void	test_2_unit (void)
{
	CrankVecFloat2	a = {3.0f, 4.0f};
	
	crank_vec_float2_unit (&a, &a);
	
	test_assert_float (a.x, 0.6f);
	test_assert_float (a.y, 0.8f);
}


void	test_2_muls (void)
{
	CrankVecFloat2	a = {3.0f, 4.0f};
	
	crank_vec_float2_muls (&a, 2.0f, &a);
	
	test_assert_float (a.x, 6.0f);
	test_assert_float (a.y, 8.0f);
}


void	test_2_divs (void)
{
	CrankVecFloat2	a = {3.0f, 4.0f};
	
	crank_vec_float2_divs (&a, 2.0f, &a);
	
	test_assert_float (a.x, 1.5f);
	test_assert_float (a.y, 2.0f);
}


void	test_2_add (void)
{
	CrankVecFloat2	a = {3.0f, 4.0f};
	CrankVecFloat2	b = {5.0f, 12.0f};
	
	crank_vec_float2_add (&a, &b, &a);
	
	test_assert_float (a.x, 8.0f);
	test_assert_float (a.y, 16.0f);
}


void	test_2_sub (void)
{
	CrankVecFloat2	a = {3.0f, 4.0f};
	CrankVecFloat2	b = {5.0f, 12.0f};
	
	crank_vec_float2_sub (&a, &b, &a);
	
	test_assert_float (a.x, -2.0f);
	test_assert_float (a.y, -8.0f);
}


void	test_2_dot (void)
{
	CrankVecFloat2	a = {3.0f, 4.0f};
	CrankVecFloat2	b = {5.0f, 12.0f};
	
	test_assert_float (crank_vec_float2_dot (&a, &b), 63.0f);
}


void	test_2_cmpmul (void)
{
	CrankVecFloat2	a = {3.0f, 4.0f};
	CrankVecFloat2	b = {5.0f, 12.0f};
	
	crank_vec_float2_cmpmul (&a, &b, &a);
	
	test_assert_float (a.x, 15.0f);
	test_assert_float (a.y, 48.0f);
}


void	test_2_cmpdiv (void)
{
	CrankVecFloat2	a = {3.0f, 4.0f};
	CrankVecFloat2	b = {5.0f, 12.0f};
	
	crank_vec_float2_cmpdiv (&b, &a, &b);
	
	test_assert_float (b.x, (5.0f / 3.0f));
	test_assert_float (b.y, 3.0f);
}


void	test_2_cmpless (void)
{
	CrankVecFloat2	a = {17.0f, 21.0f};
	CrankVecFloat2	b = {53.0f, 20.0f};
	CrankVecBool2	r;
	
	crank_vec_float2_cmpless (&a, &b, &r);
	
	g_assert_true  (r.x);
	g_assert_false (r.y);
}
	

void	test_2_cmpeq (void)
{
	CrankVecFloat2	a = {17.0f, 20.0f};
	CrankVecFloat2	b = {53.0f, 20.0f};
	CrankVecBool2	r;
	
	crank_vec_float2_cmpeq (&a, &b, &r);
	
	g_assert_false (r.x);
	g_assert_true  (r.y);
}


void	test_2_cmpgreater (void)
{
	CrankVecFloat2	a = {17.0f, 21.0f};
	CrankVecFloat2	b = {53.0f, 20.0f};
	CrankVecBool2	r;
	
	crank_vec_float2_cmpgreater (&a, &b, &r);
	
	g_assert_false (r.x);
	g_assert_true  (r.y);
}


void	test_2_cmpcmp (void)
{
	CrankVecFloat2	a = {17.0f, 21.0f};
	CrankVecFloat2	b = {53.0f, 20.0f};
	CrankVecInt2	r;
	
	crank_vec_float2_cmpcmp (&a, &b, &r);
	
	g_assert_cmpint (r.x, <, 0);
	g_assert_cmpint (r.y, >, 0);
}


void	test_2_min (void)
{
	CrankVecFloat2	a = {17.0f, 21.0f};
	CrankVecFloat2	b = {53.0f, 20.0f};
	CrankVecFloat2	r;
	
	crank_vec_float2_min (&a, &b, &r);
	
	test_assert_float (r.x, 17.0f);
	test_assert_float (r.y, 20.0f);
}


void	test_2_max (void)
{
	CrankVecFloat2	a = {17.0f, 21.0f};
	CrankVecFloat2	b = {53.0f, 20.0f};
	CrankVecFloat2	r;
	
	crank_vec_float2_max (&a, &b, &r);
	
	test_assert_float (r.x, 53.0f);
	test_assert_float (r.y, 21.0f);
}


void	test_2_mixs (void)
{
	CrankVecFloat2	a = {17.0f, 21.0f};
	CrankVecFloat2	b = {32.0f, 11.0f};
	CrankVecFloat2	r;
	
	crank_vec_float2_mixs (&a, &b, 0.2f, &r);
	
	test_assert_float (r.x, 20.0f);
	test_assert_float (r.y, 19.0f);
}


void	test_2_mix (void)
{
	CrankVecFloat2	a = {17.0f, 21.0f};
	CrankVecFloat2	b = {32.0f, 11.0f};
	CrankVecFloat2	c = {0.6f, 0.5f};
	CrankVecFloat2	r;
	
	crank_vec_float2_mix (&a, &b, &c, &r);
	
	test_assert_float (r.x, 26.0f);
	test_assert_float (r.y, 16.0f);
}



void	main (gint argc, gchar** argv)
{
	g_test_init (&argc, &argv, NULL);
	
	g_test_add_func ("/crank/base/vec/float/2/get", test_2_get);
	g_test_add_func ("/crank/base/vec/float/2/equal", test_2_equal);
	g_test_add_func ("/crank/base/vec/float/2/to_string", test_2_to_string);
	g_test_add_func ("/crank/base/vec/float/2/magn", test_2_magn);
	g_test_add_func ("/crank/base/vec/float/2/muls", test_2_muls);
	g_test_add_func ("/crank/base/vec/float/2/divs", test_2_divs);
	g_test_add_func ("/crank/base/vec/float/2/add", test_2_add);
	g_test_add_func ("/crank/base/vec/float/2/sub", test_2_sub);
	g_test_add_func ("/crank/base/vec/float/2/dot", test_2_dot);
	g_test_add_func ("/crank/base/vec/float/2/cmpmul", test_2_cmpmul);
	g_test_add_func ("/crank/base/vec/float/2/cmpdiv", test_2_cmpdiv);
	g_test_add_func ("/crank/base/vec/float/2/cmpless", test_2_cmpless);
	g_test_add_func ("/crank/base/vec/float/2/cmpeq", test_2_cmpeq);
	g_test_add_func ("/crank/base/vec/float/2/cmpgreater", test_2_cmpgreater);
	g_test_add_func ("/crank/base/vec/float/2/cmpcmp", test_2_cmpcmp);
	g_test_add_func ("/crank/base/vec/float/2/min", test_2_min);
	g_test_add_func ("/crank/base/vec/float/2/max", test_2_max);
	g_test_add_func ("/crank/base/vec/float/2/mixs", test_2_mixs);
	g_test_add_func ("/crank/base/vec/float/2/mix", test_2_mix);
	
	g_test_run ();
}

File: tests/python/test_vec_float.py
#!/usr/bin/python

# Copyright (C) 2015, WSID

# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
# THE SOFTWARE.

import unittest


from gi.repository import GObject
from gi.repository import CrankBase


class TestVecInt(unittest.TestCase):

	def assertFloat (self, a, b, delta=0.0001):
		"""A simple custom assert that given values are same.
		
		It takes into delta values into account, so that test can endure little
		errors.
		"""
		try:	#if they are both of list type.
			if (len(a) != len(b)):
				raise AssertionError ("array length: %d != %d" % (len(a), len(b)))
			
			for i in range (0, len(a)):
				if ((a[i] < b[i] - delta) or (b[i] + delta < a[i])):
					raise AssertionError ("%g != %g (diff=%g)" % (a[i], b[i], b[i]-a[i]))
		
		except TypeError: #then they are numeric type.
			if ((a < b - delta) or (b + delta < a)):
				raise AssertionError ("%g != %g (diff=%g)" % (a, b, b-a))

	def accumulation (self, value):
        	self.sum = self.sum + value;
        	return True;


	def test_2_get (self):
		a = CrankBase.VecFloat2.init (3, 4)
		
		self.assertFloat (a.get (0), 3)
		self.assertFloat (a.get (1), 4)

	def test_2_foreach (self):
		a = CrankBase.VecFloat2.init (3, 4)
		self.sum = 0;

		assert (a.foreach (self.accumulation))
		self.assertEqual (self.sum, 7)

	def test_2_equal (self):
		a = CrankBase.VecFloat2.init (3, 4)
		b = CrankBase.VecFloat2.init (5, 12)
		c = CrankBase.VecFloat2.init (3, 4)
		
		assert (not (a.equal (b)))
		assert (a.equal (c))
	
	def test_2_to_string (self):
		a = CrankBase.VecFloat2.init (3, 4)
		
		self.assertEqual (a.to_string (), "(3, 4)")
	
	def test_2_magn (self):
		a = CrankBase.VecFloat2.init (3, 4)
		
		self.assertFloat (a.get_magn_sq (), 25.0)
		self.assertFloat (a.get_magn (), 5.0)
		
	def test_2_unit (self):
		a = CrankBase.VecFloat2.init (3, 4)
		a = a.unit ()
		
		self.assertFloat (a.x, 3.0 / 5.0)
		self.assertFloat (a.y, 0.8)

	def test_2_muls (self):
		a = CrankBase.VecFloat2.init (3, 4)
		a = a.muls (3)
		
		self.assertFloat (a.x, 9)
		self.assertFloat (a.y, 12)
	
	def test_2_divs (self):
		a = CrankBase.VecFloat2.init (3, 4)
		a = a.divs (3)
		
		self.assertFloat (a.x, 1)
		self.assertFloat (a.y, 4.0 / 3.0)
				
	def test_2_add (self):
		a = CrankBase.VecFloat2.init (3, 4)
		b = CrankBase.VecFloat2.init (5, 12)
		
		a = a.add (b)
		
		self.assertFloat (a.x, 8)
		self.assertFloat (a.y, 16)
		
	def test_2_sub (self):
		a = CrankBase.VecFloat2.init (3, 4)
		b = CrankBase.VecFloat2.init (5, 12)
		
		a = a.sub (b)
		
		self.assertFloat (a.x, -2)
		self.assertFloat (a.y, -8)

	def test_2_dot (self):
		a = CrankBase.VecFloat2.init (3, 4)
		b = CrankBase.VecFloat2.init (5, 12)
		
		self.assertFloat (a.dot (b), 63)
	
	def test_2_cmpmul (self):
		a = CrankBase.VecFloat2.init (3, 4)
		b = CrankBase.VecFloat2.init (5, 12)
		
		a = a.cmpmul (b)
		
		self.assertFloat (a.x, 15)
		self.assertFloat (a.y, 48)
	
	def test_2_cmpdiv (self):
		a = CrankBase.VecFloat2.init (3, 4)
		b = CrankBase.VecFloat2.init (5, 12)
		
		b = b.cmpdiv (a)
		
		self.assertFloat (b.x, 5.0 / 3.0)
		self.assertFloat (b.y, 3)

	def test_2_cmpless (self):
		a = CrankBase.VecFloat2.init (3, 4)
		b = CrankBase.VecFloat2.init (5, 12)
		
		assert (a.cmpless (b).get_all ())
	
	def test_2_cmpeq (self):
		a = CrankBase.VecFloat2.init (3, 4)
		b = CrankBase.VecFloat2.init (3, 5)
		
		assert (a.cmpeq (b).get_any ())
	
	def test_2_cmpgreater (self):
		a = CrankBase.VecFloat2.init (3, 4)
		b = CrankBase.VecFloat2.init (5, 12)
		
		assert (not a.cmpgreater (b).get_any ())

	def test_2_cmpcmp (self):
		a = CrankBase.VecFloat2.init (3, 4)
		b = CrankBase.VecFloat2.init (3, 5)
		
		c = a.cmpcmp (b)
		
		self.assertEqual (c.x, 0)
		self.assertEqual (c.y, -1)
	
	def test_2_min (self):
		a = CrankBase.VecFloat2.init (27, 12)
		b = CrankBase.VecFloat2.init (12, 34)
		
		c = a.min (b)
		
		self.assertFloat (c.x, 12)
		self.assertFloat (c.y, 12)
		
	def test_2_max (self):
		a = CrankBase.VecFloat2.init (27, 12)
		b = CrankBase.VecFloat2.init (12, 34)
		
		c = a.max (b)
		
		self.assertFloat (c.x, 27)
		self.assertFloat (c.y, 34)
		
	
	def test_2_mixs (self):
		a = CrankBase.VecFloat2.init (15, 27)
		b = CrankBase.VecFloat2.init (30, 17)
		
		c = a.mixs (b, 0.2)
		
		self.assertFloat (c.x, 18)
		self.assertFloat (c.y, 25)
	
	def test_2_mix (self):
		a = CrankBase.VecFloat2.init (15, 27)
		b = CrankBase.VecFloat2.init (30, 17)
		c = CrankBase.VecFloat2.init (0.2, 0.8)
		
		d = a.mix (b, c)
		
		self.assertFloat (d.x, 18)
		self.assertFloat (d.y, 19)




	def test_n_get (self):
		a = CrankBase.VecFloatN.init_arr ([3, 4])
		
		self.assertFloat (a.get (0), 3)
		self.assertFloat (a.get (1), 4)

	def test_2_foreach (self):
		a = CrankBase.VecFloatN.init_arr ([3, 4])
		self.sum = 0;

		assert (a.foreach (self.accumulation))
		self.assertEqual (self.sum, 7)

	def test_n_equal (self):
		a = CrankBase.VecFloatN.init_arr ([3, 4])
		b = CrankBase.VecFloatN.init_arr ([5, 12])
		c = CrankBase.VecFloatN.init_arr ([3, 4])
		
		assert (not (a.equal (b)))
		assert (a.equal (c))
	
	def test_n_to_string (self):
		a = CrankBase.VecFloatN.init_arr ([3, 4])
		
		self.assertEqual (a.to_string (), "(3, 4)")
	
	def test_n_magn (self):
		a = CrankBase.VecFloatN.init_arr ([3, 4])
		
		self.assertFloat (a.get_magn_sq (), 25.0)
		self.assertFloat (a.get_magn (), 5.0)
		
	def test_n_unit (self):
		a = CrankBase.VecFloatN.init_arr ([3, 4])
		a = a.unit ()
		
		self.assertFloat (a.data, [3.0 / 5.0, 0.8])

	def test_n_muls (self):
		a = CrankBase.VecFloatN.init_arr ([3, 4])
		a = a.muls (3)
		
		self.assertFloat (a.data, [9, 12])
	
	def test_n_divs (self):
		a = CrankBase.VecFloatN.init_arr ([3, 4])
		a = a.divs (3)
		
		self.assertFloat (a.data, [1, 4.0 / 3.0])
				
	def test_n_add (self):
		a = CrankBase.VecFloatN.init_arr ([3, 4])
		b = CrankBase.VecFloatN.init_arr ([5, 12])
		
		a = a.add (b)
		
		self.assertFloat (a.data, [8, 16])
		
	def test_n_sub (self):
		a = CrankBase.VecFloatN.init_arr ([3, 4])
		b = CrankBase.VecFloatN.init_arr ([5, 12])
		
		a = a.sub (b)
		
		self.assertFloat (a.data, [-2, -8])

	def test_n_dot (self):
		a = CrankBase.VecFloatN.init_arr ([3, 4])
		b = CrankBase.VecFloatN.init_arr ([5, 12])
		
		self.assertFloat (a.dot (b), 63)
	
	def test_n_cmpmul (self):
		a = CrankBase.VecFloatN.init_arr ([3, 4])
		b = CrankBase.VecFloatN.init_arr ([5, 12])
		
		a = a.cmpmul (b)
		
		self.assertFloat (a.data, [15, 48])
	
	def test_n_cmpdiv (self):
		a = CrankBase.VecFloatN.init_arr ([3, 4])
		b = CrankBase.VecFloatN.init_arr ([5, 12])
		
		b = b.cmpdiv (a)
		
		self.assertFloat (b.data, [5.0 / 3.0, 3])

	def test_n_cmpless (self):
		a = CrankBase.VecFloatN.init_arr ([3, 4])
		b = CrankBase.VecFloatN.init_arr ([5, 12])
		
		assert (a.cmpless (b).get_all ())
	
	def test_n_cmpeq (self):
		a = CrankBase.VecFloatN.init_arr ([3, 4])
		b = CrankBase.VecFloatN.init_arr ([3, 5])
		
		assert (a.cmpeq (b).get_any ())
	
	def test_n_cmpgreater (self):
		a = CrankBase.VecFloatN.init_arr ([3, 4])
		b = CrankBase.VecFloatN.init_arr ([5, 12])
		
		assert (not a.cmpgreater (b).get_any ())

	def test_n_cmpcmp (self):
		a = CrankBase.VecFloatN.init_arr ([3, 4])
		b = CrankBase.VecFloatN.init_arr ([3, 5])
		
		c = a.cmpcmp (b)
		
		self.assertEqual (c.data, [0, -1])
	
	def test_n_min (self):
		a = CrankBase.VecFloatN.init_arr ([27, 12])
		b = CrankBase.VecFloatN.init_arr ([12, 34])
		
		c = a.min (b)
		
		self.assertFloat (c.data, [12, 12])
		
	def test_n_max (self):
		a = CrankBase.VecFloatN.init_arr ([27, 12])
		b = CrankBase.VecFloatN.init_arr ([12, 34])
		
		c = a.max (b)
		
		self.assertFloat (c.data, [27, 34])
		
	
	def test_n_mixs (self):
		a = CrankBase.VecFloatN.init_arr ([15, 27])
		b = CrankBase.VecFloatN.init_arr ([30, 17])
		
		c = a.mixs (b, 0.2)
		
		self.assertFloat (c.data, [18, 25])
	
	def test_n_mix (self):
		a = CrankBase.VecFloatN.init_arr ([15, 27])
		b = CrankBase.VecFloatN.init_arr ([30, 17])
		c = CrankBase.VecFloatN.init_arr ([0.2, 0.8])
		
		d = a.mix (b, c)
		
		self.assertFloat (d.data, [18, 19])
		
if __name__ == '__main__':
	unittest.main ()
